### DISCUSSION TITLE
Slow start mc(Midnight Commander) on bash with omp version >= 23.8.0

### DISCUSSION THREAD

**Discussion by aleksandr-shevchenko** [JanDeDobbeleer/oh-my-posh#5621] [Troubleshoot]
### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

### What happened?

After upgrading to 23.13.1 on varyous Linux mc(Midnight Commander) start very slow, about 20sec.
Also mc not able setting own subshell(pressing Ctrl-O).
Temporary solution for fast start - `alias mc="mc --nosubshell"`.

But on version 21.4.0 all work fine...

### Theme

custom

### What OS are you seeing the problem on?

Linux

### Which shell are you using?

bash

### Log output

```Shell
-
```

**Comment by JanDeDobbeleer** [comment]
@aleksandr-shevchenko bash (cygwin and ubuntu) start instantly on my end. There's also no way it can take 20 seconds as there's no blocking commands (except for upgrade/notice which are capped at 5 seconds). If the `--nosubshell` switch solves it, my gut feeling tells me it's either the upgrade notice, or auto upgrade that fails to correctly execute here. Which is still an MC problem to solve, it's not doing anything magical, just runs the oh-my-posh CLI on eval.

**Comment by aleksandr-shevchenko** [comment]
We are not talking about bash, everything is fine with it.

Launching Midnight Commander on the latest version of oh-my-posh in the terminal takes a very long time, up to 20 seconds and cannot connect its subshell. The fact that I disable the subshell itself simply allows MC to start instantly, but I lose some of the functionality of MC.

In the old version of OMP 21.4.0 there were no problems with the delay and MC worked correctly.

You have broked something up in the one of the latest updates. I would be grateful if you fixed the error.

**Comment by aleksandr-shevchenko** [comment]
I downloaded a dozen releases of OMP and found that everything worked correctly up to and including version 23.7.2.
In version 23.8.0 you already broke something...

**Comment by JanDeDobbeleer** [comment]
@aleksandr-shevchenko I don't mind looking into what could cause this, but please mind the language. It comes over as very aggressive, which probably isn't the intent but definitely is perceived as such on my end.

The only things that changed can be found in the release changelog, I only find the addition of the `-r` flag on the `read` command. Which we don't execute in MC (that's the only MC specific flag, we don't really support it otherwise).

I want to know if setting `disable_notice` to `true` in the configuration solves the issue, that's to my knowledge the only subshell. And if that's true, this is an MC limitation, not an oh-my-posh bug.

**Comment by aleksandr-shevchenko** [comment]
I use Google Translate. There is no aggression.

> The only things that changed can be found in the release changelog...

![image](https://github.com/user-attachments/assets/32d29a7d-6215-4f1e-aead-91d888a46da8)

**Reply by JanDeDobbeleer** [reply]
@aleksandr-shevchenko it would be great if you could try this and let me know the result though:

> I want to know if setting disable_notice to true in the configuration solves the issue, that's to my knowledge the only subshell. And if that's true, this is an MC limitation, not an oh-my-posh bug.

**Comment by aleksandr-shevchenko** [comment]
This setting solved the slow start of mc. Thanks.

PS. But I stand by my opinion. Your change in 23.8.0 has messed something up, because before it everything worked fine.

**Reply by JanDeDobbeleer** [reply]
@aleksandr-shevchenko this is oh-my-posh starting a subshell. Which should be supported by MC. It's MC that fails to start this correctly, please create a bug on their end.

**Comment by firebull** [comment]
Confirm, `disable_notice to true` solves the issue. MC starts fast and even subshell works fine.

**Comment by zyv** [comment]
@JanDeDobbeleer hi, mc maintainer here.

We must start a subshell when we boot (if the user wants to have a subshell), and due to the limitations of our current architecture, we communicate with the subshell via pipes, injecting our own PROMPT_COMMAND to get the information back from the subshell, and we have to wait until the subshell is responsive before we continue.

(By current architecture, I mean that 1) we don't have a built-in full virtual terminal emulator like screen or tmux, and 2) we don't have a proper event loop - both of which, unfortunately, are unlikely to change unless a complete rewrite is undertaken one day).

In the past, we had a relatively short timeout (2 seconds) for the subshell to start and become operational, but as projects like oh-my-posh, oh-my-zsh, etc. began to gain traction, our users started complaining that mc would randomly start without a subshell. We increased the timeout to 5-10 seconds at first, and now I think to 20 or 30 seconds (have to look at the code), and now the subshells start reliably again, but users complain that the startup is slow.

Obviously, with my mc maintainer hat on, I think it's unacceptable to have such slow startup times, no matter how much good is going on, but I guess all the other users don't care too much because the terminals are so slow to start that they mask that startup time for them. However, I'm open to working with you if there's something not so horrible we can do on our end to solve the problem for users. I understand that this crazy jump in startup time was introduced by some network update checking code that can be turned off by the user, but is on by default.

When we fork the subshell, we set the `MC_SID` environment variable. Could you please disable the heavy stuff when `MC_SID` is set? You've already made some changes to #3415 to avoid breaking our hooks depending on `MC_SID`, so you're already using it, this should only be a few character change for you, and will help us and the users a lot.

Thank you!

P.S. Of course, alternatively, we could set some other oh-my-posh specific environment variable when we fork bash, if you have a mechanism to turn it off for us externally, but I'd much rather you do that, since we can't know in advance if the bash user has oh-my-posh or not, and we'll have to set the "garbage" variable for everyone.

**Reply by zyv** [reply]
TLDR; could you please make `cfg.DisableNotice` true if `os.Getenv("MC_SID")` is set?

**Reply by JanDeDobbeleer** [reply]
@zyv check, I was onboard with that change 😃
